Web UI/CognitiveSearch.UI/wwwroot/js/results.js

```diff
@@ -87,33 +86,110 @@ function AddMapPoints(results) {
+            resultsMap.setCamera({ center: coordinates });
         }
     }
     else {
         //Create a data source to add it to the map 
-        mapDataSource = new atlas.source.DataSource();
+        mapDataSource = new atlas.source.DataSource(null, {
+            cluster: true,
+            clusterRadius: 45,
+            clusterMaxZoom: 15
+        });
         coordinates = UpdatePOIs(results, mapDataSource);
-        
+
         //Wait until the map resources are ready for first set up.
         resultsMap.events.add('ready', function () {
+            resultsMap.imageSprite.add('bubble-icon', '../images/single_bubble.png');
 
             //take the last coordinates.
-            if (coordinates) { resultsMap.setCamera ({ center: coordinates }); }
+            if (coordinates) { resultsMap.setCamera({ center: coordinates }); }
 
             //Add data source and create a symbol layer.
             resultsMap.sources.add(mapDataSource);
-            var symbolLayer = new atlas.layer.SymbolLayer(mapDataSource);
-            resultsMap.layers.add(symbolLayer);
+
+            //Create a bubble layer for rendering clustered data points.
+            var clusterBubbleLayer = new atlas.layer.BubbleLayer(mapDataSource, null, {
+                opacity: 0.7,
+                radius: [
+                    'interpolate',
+                    ['linear'],
+                    ['get', 'point_count'],
+                    2, 10,
+                    50, 35,
+                ],
+
+                color: [
+                    'interpolate',
+                    ['linear'],
+                    ['get', 'point_count'],
+                    0, 'rgba(255, 108, 92, 1)',
+                    50, 'rgba(255, 108, 92, 1)',
+                    100, 'rgba(255, 108, 92, 1)'
+                ],
+
+                strokeWidth: 0,
+                filter: ['has', 'point_count']
+            });
+
+            var symbolLayer = new atlas.layer.SymbolLayer(mapDataSource, null, {
+                iconOptions: {
+                    image: 'bubble-icon',
+                    size: 0.3
+                },
+                textOptions: {
+                    //textField: ['concat', ['to-string', ['get', 'temperature']], 'Pin'],
+                    offset: [0, -.33]
+                },
+                filter: ['!', ['has', 'point_count']]
+            });
+
+            var symbolLayer2 = new atlas.layer.SymbolLayer(mapDataSource, null, {
+                iconOptions: {
+                    image: 'none'
+                }
+            });
+
+            resultsMap.layers.add([
+                clusterBubbleLayer,
+                symbolLayer2,
+                symbolLayer
+            ]);
+
+            clusterPopup = new atlas.Popup({
+                pixelOffset: [0, 0],
+                closeButton: false
+            });
+
+            resultsMap.events.add('mouseenter', clusterBubbleLayer, function (e) {
```

Comment:
similar comment, do you think we can consolidate these events methods since they just vary in name of the event and setting different opacities?

Reply:
Do you mean create an event handler? Possibly.  The opacity setting was a work around.  Im wondering if listing all the data values via hover on/off will be acceptable.  If a single data element is preferred this would all change. (Side note: I burned a lot of hours last night trying to get  a single element response, partly due to the fact that clusters apparently do not have properties.getCoordinates.)